File: src/ast/access.cpp
#include "access.h"
#include "ast.h"
#include "call.h"
#include "literals.h"
#include "mod.h"
#include <fmt/core.h>
#include <llvm/IR/Constants.h>
#include <llvm/IR/DerivedTypes.h>
#include <llvm/IR/Value.h>
#include <tuple>
#include <vector>

namespace lygos {
    namespace AST {
        MemberExpr::MemberExpr(Ref<AST> obj, Ref<AST> member, bool deref):
            AST(ASTType::MemberExpr), obj(obj), member(member), deref(deref) {

        }

        MemberExpr::MemberExpr(Ref<AST> obj, u32 index):
            AST(ASTType::MemberExpr), obj(obj), member(nullptr), deref(false), use_index(true), index(index) {

        }

        std::string MemberExpr::GetValue() {
            return obj->GetValue();
        }

        llvm::Value *MemberExpr::GenCode(Scope *scope) {
            auto obj = this->obj->GenCode(scope);
            if(deref)
                obj = LoadOrIgnore(obj);

            //FIXME: function chain calling
            //check if function is actually a member and not static
            if(!use_index && member->type == ASTType::CallExpr) {
                std::string struct_name = static_cast<llvm::StructType *>(TryGetPointerBase(obj->getType()))->getName().data();
                auto call = (CallExpr *)member.get();
                switch (call->GetCaller()->type) {
                    case ASTType::Id: {
                        std::string &fn_name = static_cast<Identifier *>(call->GetCaller().get())->GetId();
                        fn_name = struct_name + "_" + fn_name;
                    } break;
                    default:
                        LYGOS_ASSERT(false && "?????");
                    break;
                }
                return member->GenCode(scope);
            }

            size_t index = 100000;
            std::string type_name = static_cast<llvm::StructType *>(TryGetPointerBase(obj->getType()))->getName().data();
            auto struct_fields = scope->GetStruct(type_name).fields;
            if(!use_index) {
                std::string member_name = member->GetValue();
                bool contains = false;
                for(size_t i = 0; i < struct_fields.size(); i++) {
                    if(std::get<0>(struct_fields[i]) == member_name) {
                        contains = true;
                        break;
                    }
                }
                if(!contains)
                    Log::Logger::Warn(fmt::format("unknown field `{}` in struct `{}`", member_name, type_name));
                for(size_t i = 0; i < struct_fields.size(); i++)
                    if(std::get<0>(struct_fields[i]) == member_name)
                       index = i;
            }else {
                index = this->index;
                LYGOS_ASSERT(index <= struct_fields.size() && "out of bounds index in strcut gep");
            }

            if(!use_index && member->type != ASTType::Id)
                member->GenCode(scope);

            return builder->CreateStructGEP(TryGetPointerBase(obj->getType()), obj, index);
        }

        Ref<Type::Type> MemberExpr::GetType(Scope *scope) {
            auto type = obj->GetType(scope);
            auto struct_fields = scope->GetStruct(type->GetName()).fields;
            size_t index = 0;
            for(size_t i = 0; i < struct_fields.size(); i++)
                if(std::get<0>(struct_fields[i]) == member->GetValue())
                    index = i;
            if(member->type != ASTType::Id)
                return member->GetType(scope);
            return std::get<1>(struct_fields.at(index));
        }

        void MemberExpr::Lower(AST *parent) {
            obj->Lower(this);

            if(!use_index) {
                if(member->type == ASTType::CallExpr) {
                    auto call = (CallExpr *)member.get();
                    call->Args().insert(call->Args().cbegin(), this->obj);
                }

                member->Lower(this);
            }
        }

        void MemberExpr::Sanatize() {

        }

        AccessExpr::AccessExpr(Ref<AST> obj, Ref<AST> index):
            AST(ASTType::AccessExpr), obj(obj), index(index) {

        }

        std::string AccessExpr::GetValue() {
            return obj->GetValue();
        }

        llvm::Value *AccessExpr::GenCode(Scope *scope) {
            auto obj = this->obj->GenCode(scope);
            if(!IsArrayType(obj->getType()))
                obj = LoadOrIgnore(obj);

            /*if(obj->getType()->isStructTy()) {
                auto struct_name = ((llvm::StructType *)obj->getType())->getName().data();
                if(scope->GetStruct(struct_name).ImplementsTrait("Index")) {
                    auto iden = MakeRef<Identifier>("index");
                    std::vector<Ref<AST>> args = {index};
                    auto expr = MemberExpr{this->obj, MakeRef<CallExpr>(iden, args), true};
                    return expr.GenCode(scope);
                }
                Log::Logger::Warn(fmt::format("type `{}` does not implement `Index`", struct_name));
            }*/

            auto index = this->index->GenCode(scope);
            if(ShouldLoad(this->index.get()))
                index = LoadOrIgnore(index);

            auto *zero = llvm::ConstantInt::get(llvm::Type::getInt32PtrTy(*ctx), 0);

            llvm::ArrayRef<llvm::Value *> idx_list = {zero, index};
            if(!IsArrayType(obj->getType()))
                idx_list = idx_list.drop_front();

            return builder->CreateGEP(TryGetPointerBase(obj->getType()), obj, idx_list, "", true);
        }

        Ref<Type::Type> AccessExpr::GetType(Scope *scope) {
            return obj->GetType(scope);
        }

        void AccessExpr::Lower(AST *parent) {

        }

        void AccessExpr::Sanatize() {

        }

        ResolutionExpr::ResolutionExpr(Ref<AST> obj, Ref<AST> member):
            AST(ASTType::ResolutionExpr), obj(obj), member(member) {

        }

        std::string ResolutionExpr::GetValue() {
            return obj->GetValue();
        }

        //check for static -> cant call static functions on an object
        llvm::Value *ResolutionExpr::GenCode(Scope *scope) {
            if (scope->IsEnum(obj->GetValue())) {
                Type::EnumType e = scope->GetEnum(obj->GetValue());
                for(size_t i = 0; i < e.variants.size(); i++) {
                    if(e.variants[i] == this->member->GetValue()){
                        return llvm::ConstantInt::get(scope->GetType(e.type.get()), i);
                    }
                }
                Log::Logger::Warn(fmt::format("unknown enum variant `{}` in enum `{}`", member->GetValue(), e.name));
            }

            auto fn = (CallExpr *)member.get();
            ((Identifier *)fn->GetCaller().get())->GetId() = scope->GetStruct(obj->GetValue()).GetFunction(fn->GetCaller()->GetValue()).name_mangeled;
            return member->GenCode(scope);
        }

        Ref<Type::Type> ResolutionExpr::GetType(Scope *scope) {
            return member->GetType(scope);
        }

        void ResolutionExpr::Lower(AST *parent) {

        }

        void ResolutionExpr::Sanatize() {

        }
    }
}

File: src/ast/access.rs
use crate::ast::{AST, Generate};
use crate::types::{Type, Path};
use crate::lexer::Tagged;
use crate::log::{ErrorLabel, error_msg_label_info, error_msg_label};

#[derive(Debug)]
pub struct MemberExpr {
    obj: Box<AST>,
    pub member: Option<Box<AST>>,
    deref: bool,
    use_index: bool,
    index: usize,
}

impl MemberExpr {
    pub fn new(obj: Box<AST>, member: Box<AST>, deref: bool) -> Self {
        Self {
            obj,
            member: Some(member),
            deref,
            use_index: false,
            index: 0,
        }
    }

    pub fn new_index(obj: Box<AST>, index: usize) -> Self {
        Self {
            obj,
            member: None,
            deref: false,
            use_index: true,
            index,
        }
    }
}

impl Generate for MemberExpr {
    fn loc(&self) -> &crate::lexer::Loc {
        self.obj.loc()
    }

    fn get_value(&self) -> String {
        self.obj.get_value()
    }

    fn gen_code(&mut self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<llvm::ValueRef> {
        let mut obj = self.obj.gen_code(scope, ctx).unwrap();
        let obj_ty = self.obj.get_type(scope, ctx).unwrap();
        if self.deref {
            if let Some(base) = obj_ty.get_base() {
                obj = obj.try_load(&scope.resolve_type(&base, ctx), ctx.builder);
            }
        }

        if let Type::Pointer(ptr) = &obj_ty {
            if ptr.is_ref {
                let base = &ptr.typ;
                obj = obj.try_load(&scope.resolve_type(base, ctx), ctx.builder);
            }
        }
        let type_name = obj_ty.get_name();
        let struct_fields = scope.get_struct(&Tagged::new(self.obj.loc().clone(), type_name.clone()));
        let index = if !self.use_index {
            let mut index_internal = None;
            let member_name = self.member.as_ref().unwrap().get_value();
            for (i, field) in struct_fields.fields.iter().enumerate() {
                if field.id.inner().clone() == member_name {
                    index_internal = Some(i);
                }
            }

            if index_internal.is_none() {
                let mut fields_avail = "available fieds: [".to_string();
                for (i, field) in struct_fields.fields.iter().enumerate() {
                    fields_avail += &("`".to_string() + field.id.inner() + "`");
                    if i + 1 != struct_fields.fields.len() {
                        fields_avail += ", ";
                    }
                }
                fields_avail += "]";
                error_msg_label_info(
                    format!("unknown field `{}` in type `{}`", member_name, type_name).as_str(),
                    ErrorLabel::from(self.member.as_ref().unwrap().loc(), "unknown field"),
                    &fields_avail
                );
            }
            index_internal.unwrap()
        }else {
            self.index
        };

        if let Some(mem) = &self.member {
            if let AST::Id(_) = **mem {} else {
                if !self.use_index {
                    self.member.as_mut().unwrap().gen_code(scope, ctx);
                }
            }
        }

        if !obj.get_type().is_pointer_ty() {
            let alloc = ctx.builder.create_alloca(&obj.get_type(), None);
            ctx.builder.create_store(&obj, &alloc);
            obj = alloc;
        }

        let base: Option<Type> = if self.deref {
            let tmp = self.obj.get_type(scope, ctx).unwrap();
            tmp.get_base().cloned()
        }else if let Type::Pointer(ptr) = &obj_ty {
            if ptr.is_ref {
                Some(*ptr.typ.clone())
            }else {
                Some(obj_ty)
            }
        }else {
            Some(obj_ty)
        };
        if let Some(ty) = base {
            if ty.get_base().is_some() {
                /*
                *  FIXME(S): provide a better error & help message this does not cut it XD
                *
                */
                error_msg_label_info(
                    "invalid level of indirection",
                    ErrorLabel::from(self.loc(), &format!("cannot access member of type `{}`", self.obj.get_type(scope, ctx).unwrap().get_full_name())),
                    "try dereferencing using `->` instead of `.`"
                );
            }
            return Some(ctx.builder.create_struct_gep(&scope.resolve_type(&ty, ctx), &obj, index as u32));
        }

        error_msg_label(
            "cannot deref value type",
            ErrorLabel::from(self.loc(), format!("cannot deref value type `{}`", self.obj.get_type(scope, ctx).unwrap().get_full_name()).as_str())
        );
    }

    fn get_type(&self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<crate::types::Type> {
        let obj_ty = self.obj.get_type(scope, ctx).unwrap();
        let struct_fields = &scope.get_struct(&Tagged::new(self.obj.loc().clone(), obj_ty.get_name())).fields;
        let index = if !self.use_index {
            let mut index_internal = None;
            let member_name = self.member.as_ref().unwrap().get_value();
            for (i, field) in struct_fields.iter().enumerate() {
                if field.id.inner().clone() == member_name {
                    index_internal = Some(i);
                }
            }

            if index_internal.is_none() {
                let mut fields_avail = "available fieds: [".to_string();
                for (i, field) in struct_fields.iter().enumerate() {
                    fields_avail += &("`".to_string() + field.id.inner() + "`");
                    if i + 1 != struct_fields.len() {
                        fields_avail += ", ";
                    }
                }
                fields_avail += "]";
                error_msg_label_info(
                    format!("unknown field `{}` in type `{}`", member_name, obj_ty.get_name()).as_str(),
                    ErrorLabel::from(self.member.as_ref().unwrap().loc(), "unknown field"),
                    &fields_avail
                );
            }
            index_internal.unwrap()
        }else {
            self.index
        };
        return Some(struct_fields[index].typ.clone());
    }

    fn collect_symbols(&mut self, _scope: &mut super::Scope) {}
}

#[derive(Debug)]
pub struct AccessExpr {
    obj: Box<AST>,
    index: Box<AST>,
}

impl AccessExpr {
    pub fn new(obj: Box<AST>, index: Box<AST>) -> Self {
        Self{ obj, index }
    }
}

impl Generate for AccessExpr {
    fn loc(&self) -> &crate::lexer::Loc {
        self.obj.loc()
    }

    fn get_value(&self) -> String {
        self.obj.get_value()
    }

    fn gen_code(&mut self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<llvm::ValueRef> {
        let mut obj = self.obj.gen_code(scope, ctx).unwrap();
        /*if !crate::types::is_array_type(&obj.get_type()) {
            obj = obj.try_load(ctx.builder);
        }*/
        /*if let Type::Array(arr) = self.obj.get_type(scope, ctx).unwrap() {
            obj = obj.try_load(&scope.resolve_type(&arr.typ, ctx), ctx.builder);
        }*/

        let mut index = self.index.gen_code(scope, ctx).unwrap();
        if self.index.should_load() {
            let base = self.index.get_type(scope, ctx).unwrap();
            index = index.try_load(&scope.resolve_type(&base, ctx), ctx.builder);
        }

        let zero = llvm::ConstantInt::get(&llvm::TypeRef::get_int(ctx.ctx, 64), 0);

        let mut idx_list = vec![index];
        if let Type::Array(_) = self.obj.get_type(scope, ctx).unwrap() {
            idx_list.insert(0, zero);
        }

        //if let Some(ty) = self.obj.get_type(scope, ctx).unwrap().get_base() {
        let base = self.obj.get_type(scope, ctx).unwrap();
        return Some(ctx.builder.create_gep(&scope.resolve_type(&base, ctx), &obj, &idx_list, true));
        //}

        //error_msg_label(
        //    "cannot deref value type",
        //    ErrorLabel::from(self.loc(), format!("cannot deref value type `{}`", self.obj.get_type(scope, ctx).unwrap().get_full_name()).as_str())
        //);
    }

    fn get_type(&self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<crate::types::Type> {
        match self.obj.get_type(scope, ctx).unwrap() {
            Type::Array(arr) => return Some(*arr.typ),
            Type::Pointer(ptr) => return Some(*ptr.typ),
            ty => return Some(ty),
        }
    }

    fn collect_symbols(&mut self, _scope: &mut super::Scope) {}
}

#[derive(Debug)]
pub struct ResolutionExpr {
    obj: Box<AST>,
    member: Box<AST>,
}

impl ResolutionExpr {
    pub fn new(obj: Box<AST>, member: Box<AST>) -> Self {
        Self{ obj, member }
    }
}

impl Generate for ResolutionExpr {
    fn loc(&self) -> &crate::lexer::Loc {
        self.obj.loc()
    }

    fn get_value(&self) -> String {
        self.obj.get_value()
    }

    fn gen_code(&mut self, scope: &mut super::Scope, ctx: &crate::GenerationContext) -> Option<llvm::ValueRef> {
        let member_value = self.member.get_value();
        let tagged = Tagged::new(self.obj.loc().clone(), self.obj.get_value());
        if scope.is_enum(&tagged) {
            let ev = scope.get_enum(&tagged);
            for (i, var) in ev.variants.iter().enumerate() {
                if *var == member_value {
                    return Some(llvm::ConstantInt::get(&scope.resolve_type(&ev.typ, ctx), i as i32));
                }
            }
            error_msg_label(
                format!("unknown enum variant `{}` in enum `{}`", member_value, tagged.inner()).as_str(),
                ErrorLabel::from(self.member.loc(), "unknown enum variant"),
            );
        }

        if let AST::CallExpr(call) = &mut *self.member {
            return call.gen_code_internal(scope, ctx, Some((&self.obj, true)), None);
        }

        //provide an actual error essage
        error_msg_label(
            format!("resolution expr err").as_str(),
            ErrorLabel::from(self.member.loc(), "unknown resolution expr"),
        );
    }

    fn get_type(&self, scope: &mut super::Scope, _: &crate::GenerationContext) -> Option<crate::types::Type> {
        let tagged = Tagged::new(self.obj.loc().clone(), self.obj.get_value());
        if scope.is_enum(&tagged) {
            return Some(Type::Path(Path::new(self.obj.loc().clone(), self.obj.get_value())));
        }
        let strct = scope.get_struct(&tagged);
        let func = strct.get_function(self.member.loc(), self.member.get_value());
        return Some(func.ret_type.clone());
    }

    fn collect_symbols(&mut self, _scope: &mut super::Scope) {}
}
